cmd/dep/ensure.go

```diff
@@ -731,18 +737,46 @@ func checkErrors(m map[string]pkgtree.PackageOrErr) error {
 			case *build.NoGoError:
 				noGoErrors++
 			default:
-				buildErrors = append(buildErrors, poe.Err.Error())
+				pkgtreeErrors.errs = append(pkgtreeErrors.errs, poe.Err)
 			}
 		}
 	}
 
+	// If pkgtree was empty or all dirs lacked any Go code, return an error.
 	if len(m) == 0 || len(m) == noGoErrors {
-		return errors.New("all dirs lacked any go code")
+		return errors.New("no dirs contained any Go code")
+	}
+
+	// If all dirs contained build errors, return an error.
+	if len(m) == len(pkgtreeErrors.errs) {
+		return errors.New("all dirs contained build errors")
+	}
+
+	// If all directories either had no Go files or caused a build error, return an error.
+	if len(m) == len(pkgtreeErrors.errs)+noGoErrors {
+		return pkgtreeErrors
 	}
 
-	if len(buildErrors) > 0 {
-		return errors.Errorf("Found %d errors:\n\n%s", len(buildErrors), strings.Join(buildErrors, "\n"))
+	// If m contained some errors, return a warning with these errors.
+	if len(pkgtreeErrors.errs) > 0 {
+		pkgtreeErrors.warn = true
+		return pkgtreeErrors
 	}
 
 	return nil
 }
+
+type pkgtreeErrs struct {
+	errs []error // errs is a list of build errors
+	warn bool    // warn determines if this error should be treated as a warning or an error.
+}
+
+func (e *pkgtreeErrs) Error() string {
+	var errors []string
```

Comment:
nit: could init slice to exact length here: `errors := make([]string, 0, len(e.errs))` (or only make with a length, and use the index instead of append in the loop)